MatrixKit/Views/Authentication/MXKAuthenticationFallbackWebView.m

```diff
@@ -119,24 +119,21 @@ - (BOOL)webView:(UIWebView *)webView shouldStartLoadWithRequest:(NSURLRequest *)
             if ([@"onRegistered" isEqualToString:parameters[@"action"]])
             {
                 // Translate the JS registration event to MXCredentials
+                // We cannot use [MXCredentials modelFromJSON:] because of https://github.com/matrix-org/synapse/issues/4756
+                // Because of this issue, we cannot get the device_id allocated by the homeserver
+                // TODO: Fix it once the homeserver issue is fixed (filed at https://github.com/vector-im/riot-meta/issues/273).
                 MXCredentials *credentials = [[MXCredentials alloc] initWithHomeServer:parameters[@"homeServer"] userId:parameters[@"userId"] accessToken:parameters[@"accessToken"]];
+
                 // And inform the client
                 onSuccess(credentials);
             }
             else if ([@"onLogin" isEqualToString:parameters[@"action"]])
             {
                 // Translate the JS login event to MXCredentials
-                NSString *homeServer = parameters[@"response"][@"home_server"];
-                NSString *userId = parameters[@"response"][@"user_id"];
-                NSString *accessToken = parameters[@"response"][@"access_token"];
-                
-                // Sanity check
-                if (homeServer.length && userId.length && accessToken.length)
-                {
-                    MXCredentials *credentials = [[MXCredentials alloc] initWithHomeServer:homeServer userId:userId accessToken:accessToken];
-                    // And inform the client
-                    onSuccess(credentials);
-                }                
+                MXCredentials *credentials = [MXCredentials modelFromJSON:parameters[@"response"]];
```

Comment:
If `parameters[@"response"]` is nil an empty MXCredentials will be created. We could maybe add a failure closure to this method `- (void)openFallbackPage:(NSString*)fallbackPage success:(void (^)(MXCredentials *credentials))success;`

Reply:
This is a PR for a fix for device_id not a PR for a behavior change.